framework/include/convergence/Convergence.h

```diff
@@ -0,0 +1,52 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "MooseObject.h"
+#include "SetupInterface.h"
+#include "PostprocessorInterface.h"
+#include "PerfGraphInterface.h"
+
+/**
+ * Base class for convergence criteria.
+ */
+class Convergence : public MooseObject,
+                    public SetupInterface,
+                    public PostprocessorInterface,
+                    public PerfGraphInterface
+{
+public:
+  static InputParameters validParams();
+
+  /**
+   * Status returned by calls to \c checkConvergence.
+   */
+  enum class MooseConvergenceStatus
+  {
+    ITERATING = 0,
+    CONVERGED = 2,
+    DIVERGED = -2
+  };
+
+  Convergence(const InputParameters & parameters);
+
+  virtual void initialSetup() override{};
+
+  /**
+   * Returns convergence status.
+   *
+   * @param[in] iter   Iteration index
+   */
+  virtual MooseConvergenceStatus checkConvergence(unsigned int iter) = 0;
```

Comment:
Yeah this relates to one of my discussions in the PR description. I'm trying to figure out what the best design will be when it comes time to extend the system to include other iteration types (like fixed point or steady-state detection). How might we use one generic Convergence for multiple iteration types?

Reply:
It d be cool to be able to use the same CV classes for all the iterations

Reply:
which likely precludes retrieving everything in the class

Reply:
Yeah for generic CV classes, I definitely wouldn't want a different class for each type of iteration. I'd want to somehow distinguish within `checkConvergence()` the type of iteration, whether it gets passed in as an argument like `IterationType`, or whether each *object* is associated with a single iteration type. Maybe the user tells it, or maybe it is determined internally from `nonlinear_convergence = my_conv`.

Reply:
I still don't see how this is different from other parameter types. How are you going to make sure you can use the same convergence class that checks a residual reduction for different solution schemes? E.g. for fixed point iterations you may want the L2 norm of the residual across the multiapps vs. nonlinear iterations just looking at the residual of a single app

Reply:
>  E.g. for fixed point iterations you may want the L2 norm of the residual across the multiapps

for that you use some for "MultiAppResidualFixedPointConvergence"

> nonlinear iterations just looking at the residual of a single app

for that you ideally could use the same "DefaultNonlinearConvergence" which you used on the main app

> I still don't see how this is different from other parameter types.

It is different because the concept of an iteration number is well-defined with an int for all types of iterative solves. 
This lets us use the same CV object for multiple solvers. On the other hand, the concept of a step change or a residual refers to a particular solution vector or a particular solver so you need to give more context, which you can either get from:
- implementing CV objects for specific iterations
- passing everything through checkConvergence(), but that's not sustainable
- another option I cant think of

Reply:
> whether each object is associated with a single iteration type

I think for some objects it will be inevitable. For others it should be possible, like PPCV or IterNumCV

Reply:
I'm not dead-set on having iteration as an argument. It was special because it's an input to generic CV types that I envision using for all iteration types. I think having some kind of iteration type specified is inevitable when we introduce others. We'll probably have something like
```
MultiMooseEnum validIterationTypes() const = 0;
```
and then as I'm thinking about it, probably
```
params.add(Private?)Param<MooseEnum>("iteratation_type", ...
```
There'd be a check in the constructor of `_iteration_type` vs. `validIterationTypes()`, and then in the body of `checkConvergence()`, we could do:
```
switch(_iteration_type)
{
  case Convergence::IterationType::Nonlinear:
    iter = SNESGetIteration();
  case Convergence::IterationType::SteadyState:
    iter = _problem.thisIsADumbExample();
}
```

Reply:
if we have

> MultiMooseEnum validIterationTypes() const = 0;

this would fall under

> implementing CV objects for specific iteration types

For people implementing their custom solves (e.g. Griffin's Richardson for example), they would not be able to use our CV objects

Reply:
I think I would feel better about having `it` as a parameter if you did not also have this `IterationCountConvergence` intermediate class. The latter's existence suggests to me that there is something insufficient about the base class despite the fact that there is an `it` parameter. If we are going to have this `it` parameter, I don't see why we need the intermediate class

Also, can we stop using CV to stand for convergence? Whenever I see it I just think const-volatile

Reply:
> Also, can we stop using CV to stand for convergence? Whenever I see it I just think const-volatile

man these requests 😆 

> you did not also have this IterationCountConvergence intermediate class

A good way to get rid of the intermediate class imo would be to use a "Combo" like behavior between multiple convergence classes, as is done for TimeSteppers. This would be automatically set up from all the Convergence objects specified to the "nonlinear_convergence" parameter of the FEProblem

I dont think I like that `IterationCountConvergence` is an intermediate class either. I would see it as a child class the same level as `PPConvergence`. But giving the PPConvergence one more criteria it can apply does not really hurt it either

Reply:
I think "IterationCount" should be folded into the base class. It makes sense to me since the base class interface takes the `it` parameter

Reply:
"Folded in" meaning that I think the functionality should be coded in the `Convergence` class, and that functionality can be called by child classes if they wish to incorporate iteration counting as part of their convergence criteria

Reply:
So you mean all Convergence objects have the min/max parameters? And maybe raise the max default to the numeric max to completely disable by default? I think it could work. It just might not make sense for all of them to have the parameters, like the ParsedConvergence. There's always `suppressParam` I guess.

Reply:
Discussing the iteration intermediate class on other conversation threads. Marking this as resolved